### PR TITLE
JAX reimplementation of MPNN

### PR DESCRIPTION
**Pull request by marpom03**
This PR adds a JAX reimplementation of the MPNN model.

### CODE REVIEW

File: mpnn/mpnn_jax/loader.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    W  = jnp.asarray(W)
+
+    mu_norm = (mu - norm.mu_shift) / norm.mu_scale
+    W_norm  = (W  - norm.W_shift) / norm.W_scale
+
+    nodes = jnp.concatenate([mu_norm, L], axis=1)                
+    senders, receivers = jnp.where(R > 0)
+    edges = R[senders, receivers].reshape(-1, 1)
+
+    return jraph.GraphsTuple(
```
Comment (krzysztofrusek) [inline mpnn/mpnn_jax/loader.py:34]: Graph should be padded to avoid `jit` recompilation.
I recommend paddling to the nearest power of two.
For training, loss must be masked and computed only for non-padded graph

File: mpnn/mpnn_jax/model_nnx/model.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        i_out = self.i2(i_out)
+
+        j_out = jnn.selu(self.j1(hx))
+        j_out = self.j2(j_out)
+
+        RR = jnn.sigmoid(i_out) * j_out  # (N, rn)
+        if node_mask is not None:
+            RR = RR * node_mask.astype(RR.dtype)
+
+        graph_idx = _graph_indices_from_n_node(graph)
```
Comment (krzysztofrusek) [inline mpnn/mpnn_jax/model_nnx/model.py:140]: Mask padded nodes

File: mpnn/mpnn_jax/train_linen.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+)
+
+
+@jax.jit
+def loss_fn(params, batch):
+    """
+    Compute the MSE loss.
+    """
+    preds = model.apply(params, batch)                 
+    labels = batch[0].globals.squeeze()
```
Comment (krzysztofrusek) [inline mpnn/mpnn_jax/train_linen.py]: why `batch[0]` ?

File: mpnn/mpnn_jax/train_linen.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+
+@jax.jit
+def loss_fn(params, batch):
+    """
+    Compute the MSE loss.
+    """
+    preds = model.apply(params, batch)                 
+    labels = batch[0].globals.squeeze()                
+    loss = jnp.mean((preds - labels)**2)
```
Comment (krzysztofrusek) [inline mpnn/mpnn_jax/train_linen.py:30]: Compute only for real graph and skip padding

File: mpnn/mpnn_jax/train_nnx.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        model = nnx.merge(graphdef, full_state)
+        return model(batch)  # batch = (graph, node_mask, edge_mask)
+    return apply_fn
+
+def build_loss_fn(apply_fn):    
+    """
+    Wrap the forward pass with MSE loss in normalized label space.
+    """
+    def loss_fn(params, batch):
+        preds = apply_fn(params, batch)
```
Comment (krzysztofrusek) [inline mpnn/mpnn_jax/train_nnx.py]: SAme comments as in `linen` version

**Review by krzysztofrusek — CHANGES_REQUESTED** [review]

**Comment by marpom03** [comment]
@krzysztofrusek 
Thank you for your feedback. We have rewritten the solutions to follow the approach with padding and masks as you suggested. All the points you commented on have now been addressed and resolved.

File: mpnn/mpnn_jax/loader.py (hunk 1 above)
Comment (marpom03) [reply mpnn/mpnn_jax/loader.py:34]: The padding mechanism has now been implemented in loader.py. The loss function excludes the padded (dummy) graphs using` [:-1] ` slicing, and additional masking has been introduced inside the model to properly ignore padded edges and nodes during training. The function ` _nearest_bigger_power_of_two` was added to loader.py and model_linen/helpers.py for paddling to the nearest power of two.

File: mpnn/mpnn_jax/model_nnx/model.py (hunk 2 above)
Comment (marpom03) [reply mpnn/mpnn_jax/model_nnx/model.py:140]: Masking for padded nodes has been added.

File: mpnn/mpnn_jax/train_linen.py (hunk 3 above)
Comment (marpom03) [reply mpnn/mpnn_jax/train_linen.py]: Previously, indexing with batch[0] was required because the batch structure was (graphs, edge_mask, node_mask). Since masking is now handled differently, the correct access is batch.globals.squeeze().

File: mpnn/mpnn_jax/train_linen.py (hunk 4 above)
Comment (marpom03) [reply mpnn/mpnn_jax/train_linen.py:30]: We now use` preds[:-1]`, which ensures that only the real graph is evaluated and the padded dummy graph is skipped.

File: mpnn/mpnn_jax/train_nnx.py (hunk 5 above)
Comment (Zocha1) [reply mpnn/mpnn_jax/train_nnx.py]: The NNX version has been reimplemented and aligned with the updates made in the Linen version.

File: mpnn/mpnn_jax/plots/linen/ba_ba/eval.svg
Comment (krzysztofrusek) [inline mpnn/mpnn_jax/plots/linen/ba_ba/eval.svg:1]: Strange, the error is biased, and NNX version looks better, can takie a deeper lot at flax implementation?

**Review by krzysztofrusek — APPROVED** [review]
